src/data/attribute-fetchers.ts

```diff
@@ -0,0 +1,428 @@
+/**
+ * AR.IO Gateway
+ * Copyright (C) 2022-2023 Permanent Data Solutions, Inc. All Rights Reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import {
+  ContiguousDataSource,
+  ContiguousDataIndex,
+  ContiguousDataAttributes,
+  ChainSource,
+  TransactionAttributesStore,
+  DataItemAttributesStore,
+  DataItemAttributes,
+  TransactionAttributes,
+  B64UrlStore,
+} from '../types.js';
+import winston from 'winston';
+import { toB64Url } from '../lib/encoding.js';
+import { isEmptyString } from '../lib/string.js';
+
+export class AttributeFetchers {
+  protected log: winston.Logger;
+  protected dataSource: ContiguousDataSource;
+  protected dataIndex: ContiguousDataIndex;
+  private dataItemAttributesStore: DataItemAttributesStore;
+  private transactionAttributesStore: TransactionAttributesStore;
+
+  constructor({
+    log,
+    dataSource,
+    dataIndex,
+    dataItemAttributesStore,
+    transactionAttributesStore,
+  }: {
+    log: winston.Logger;
+    dataSource: ContiguousDataSource;
+    dataIndex: ContiguousDataIndex;
+    dataItemAttributesStore: DataItemAttributesStore;
+    transactionAttributesStore: TransactionAttributesStore;
+  }) {
+    this.log = log.child({ class: this.constructor.name });
+    this.dataSource = dataSource;
+    this.dataIndex = dataIndex;
+    this.dataItemAttributesStore = dataItemAttributesStore;
+    this.transactionAttributesStore = transactionAttributesStore;
+  }
+
+  async fetchDataFromParent({
+    parentId,
+    offset,
+    size,
+  }: {
+    parentId: string;
+    offset: number;
+    size: number;
+  }): Promise<string> {
+    const log = this.log.child({ method: 'fetchDataFromParent' });
+    log.debug('Fetching data from parent', { parentId, offset, size });
+
+    const { stream } = await this.dataSource.getData({
+      id: parentId,
+      dataAttributes: {
+        size,
+      } as ContiguousDataAttributes,
+      region: {
+        offset,
+        size,
+      },
+    });
+
+    let buffer = Buffer.alloc(0);
+
+    for await (const chunk of stream) {
+      buffer = Buffer.concat([buffer, chunk]);
+    }
+
+    return toB64Url(buffer);
+  }
+
+  async getDataItemAttributes(
+    id: string,
+  ): Promise<DataItemAttributes | undefined> {
+    const log = this.log.child({ method: 'getDataItemAttribute' });
```

Comment:
Small typo here. Should be `getDataItemAttributes` (with an 's').

Reply:
https://github.com/ar-io/ar-io-node/pull/356/commits/15e95b6af5daae18b1af19c74a1f6ced5d729a51